### PR TITLE
Replace node especific process.env with injected variables at React.

### PR DESCRIPTION
**Pull request by mshima**
Fixes https://github.com/jhipster/generator-jhipster/issues/15074.

Apply same fix used for angular https://github.com/jhipster/generator-jhipster/pull/15107. 

---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] The JDL part is updated if necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

File: generators/client/templates/react/webpack/webpack.common.js.ejs [hunk 1]
```diff
@@ -111,24 +109,13 @@ module.exports = options => merge(
   },
   plugins: [
     new webpack.DefinePlugin({
-      'process.env': {
-        NODE_ENV: `'${options.env}'`,
```
Comment (deepu105) [inline generators/client/templates/react/webpack/webpack.common.js.ejs:115]: I think React and some libs like Redux for example internally still use `process.env.NODE_ENV` for certain logic and hence we should ensure its value is passes as well IMO, unless its automatically set

File: generators/client/templates/react/webpack/environment.js.ejs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-%>
+<%_ if (buildToolUnknown) { _%>
+const packageJson = require('./../package.json');
+<%_ } _%>
+
+module.exports = {
+  __ENVIRONMENT__: 'production',
```
Comment (deepu105) [inline generators/client/templates/react/webpack/environment.js.ejs:24]: what about `development`?

**Review by deepu105 — CHANGES_REQUESTED** [review]

File: generators/client/templates/react/src/main/webapp/app/config/constants.ts.ejs [hunk 3]
```diff
@@ -16,13 +16,25 @@
...
+declare const __BUILD_TIMESTAMP__: string;
+<%_ } _%>
+declare const __VERSION__: string;
+declare const __SERVER_API_URL__: string;
+declare const __ENVIRONMENT__: string;
+
+<%_ if (enableTranslation) { _%>
+export const BUILD_TIMESTAMP = __BUILD_TIMESTAMP__;
+<%_ } _%>
+export const ENVIRONMENT = __ENVIRONMENT__;
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/client/templates/react/src/main/webapp/app/config/constants.ts.ejs:29]: > Define plugin inlines the actual code on compilation and the minification will remove the unreachable code from what I recall. Would that still happen?

Yes, but it should smarter now and remove constants too, not sure.
We could use __ENVIRONMENT__ directly too, but I think it's cleaner to have a file for inputs.

File: generators/client/templates/react/webpack/webpack.common.js.ejs (hunk 1 above)
Comment (mshima) [reply generators/client/templates/react/webpack/webpack.common.js.ejs:115]: I don't know Redux.
Use of `process.env` should be avoided: https://github.com/webpack/changelog-v5/blob/master/MIGRATION%20GUIDE.md#level-5-runtime-errors.

File: generators/client/templates/react/webpack/environment.js.ejs (hunk 2 above)
Comment (mshima) [reply generators/client/templates/react/webpack/environment.js.ejs:24]: This is used by jest, __ENVIRONMENT__ is replaced at webpack.common.js.

File: generators/client/templates/react/webpack/webpack.common.js.ejs (hunk 1 above)
Comment (deepu105) [reply generators/client/templates/react/webpack/webpack.common.js.ejs:115]: Ya but many React frameworks might use it s its better to make sure its set. Redux definitely does it for enabling dev tools and stuff

File: generators/client/templates/react/src/main/webapp/app/config/constants.ts.ejs (hunk 3 above)
[the remark this answers is not in the document]
Comment (deepu105) [reply generators/client/templates/react/src/main/webapp/app/config/constants.ts.ejs:29]: can you look at a build and verify, if its replaced its fine else would be nice to use directly so production bundles will have unwanted code removed

File: generators/client/templates/react/webpack/webpack.common.js.ejs (hunk 1 above)
Comment (mshima) [reply generators/client/templates/react/webpack/webpack.common.js.ejs:142]: According to https://webpack.js.org/guides/production/#specify-the-mode.
NODE_ENV is set by webpack when mode is set, so no need to set it manually.

File: generators/client/templates/react/src/main/webapp/app/config/constants.ts.ejs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mshima) [reply generators/client/templates/react/src/main/webapp/app/config/constants.ts.ejs]: Looks good.

**Comment by mshima** [comment]
Dropped vue changes, it doesn't have the linked issue and it needs some cleanup.

**Comment by mshima** [comment]
It's been open for too long. Merging.

**Pull request merged**